src/common/buf.rs

```diff
@@ -0,0 +1,95 @@
+use hyper::body::Buf;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use tokio::sync::Notify;
+
+/// Signaler returned as part of [`NotifyOnEos::new`] that can be polled to receive information,
+/// when the buffer gets advanced to the end.
+// Cannot be Clone due to usage of `Notify::notify_one` in `NotifyOnEos::advance`,
+// revisit once `Notify::notify_all` stabilizes.
+pub struct EosSignaler {
+    notifier: Arc<Notify>,
+}
+
+impl EosSignaler {
+    pub async fn wait_till_eos(self) {
+        self.notifier.notified().await;
+    }
+}
+
+/// Wrapper for [`Buf`] that returns an [`EosSignaler`] that can be polled to receive information,
+/// when the buffer gets advanced to the end.
+///
+/// NOTE: For the notification to work, caller must ensure that [`Buf::advance`] gets called
+/// enough times to advance to the end of the buffer (so that [`Buf::has_remaining`] afterwards returns `0`).
+pub struct NotifyOnEos<B> {
+    inner: B,
+    notifier: Arc<Notify>,
+    // It'd be better if we consumed the signaler, making it inaccessible after notification.
+    // Unfortunately, that would require something like AtomicOption.
```

Comment:
Why would `AtomicOption` be required? The signal happens in `advance`, which has `&mut self` as the receiver. So it should be possible to notify and drop it, no?

Reply:
Heh, you're right. I tried so hard to make it an `Arc<Option<...>>` never once realizing what I actually need is `Option<Arc<...>>`.

Changed in latest commit.